File: Go/Linked List/2130_Maximum Twin Sum of a Linked List.go
import "math"

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
//method1
//find mid
//reverse second half of LL
//iterate first and second to get sum

//or
//method2
//find middle
//push to stack
//traverse through LL after middle while calculating twin sum

func findMid(head *ListNode) (first, prev *ListNode) {
	if head == nil {
		return nil, nil
	}
	if head.Next == nil {
		return
	}
	first, next := head, head
	for next != nil && next.Next != nil {
		prev = first
		first = first.Next
		next = next.Next.Next
	}
	return first, prev
}

func reverseAndSum(head, mid, prev *ListNode) int {
	if mid == nil {
		return 0
	}
	cur, next := mid, mid
	for cur != nil {
		next = cur.Next
		cur.Next = prev
		prev, cur = cur, next
	}
	//prev now = head of reversed LL
	temp := head
	maxSum := math.MinInt64
	for prev.Next != temp {
		sum := temp.Val + prev.Val
		if sum > maxSum {
			maxSum = sum
		}
		temp, prev = temp.Next, prev.Next
	}
	sum := temp.Val + prev.Val
	if sum > maxSum {
		return sum
	}
	return maxSum
}

func pairSum(head *ListNode) int {
	//find mid
	mid, prev := findMid(head)
	//reverse second half of LL and find sum
	return reverseAndSum(head, mid, prev)
}

File: C++/Linked List/2130_Maximum Twin Sum of a Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        vector<int> nums{};
        while(head != nullptr){
            nums.push_back(head->val);
            head = head->next;
        }
        int beg =0,end=nums.size()-1,maxSum=nums[0];
        while(beg<end){
            int sum = nums[beg] + nums[end];
            if(sum > maxSum)
                maxSum = sum;
            beg++;
            end--;
        }
        return maxSum;
    }
};